### PR TITLE
4715: Autolink feature

### LINKED ISSUE #4715: Implement autolinking feature
Create autolinking mechanism as separate feature. After new content is inserted to the editor, editor should detect matched UTL and automatically transforms it to link.

In addition to URLs it might also support email addresses.

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (link): Added the `AutoLink` feature which replaces a plain text with a link if typed or pasted content is the link. Closes #4715.

---

### Additional information

*   I'm not sure how verbose the feature guide should be.
*   I've might forget something due to various distractions so sorry if found something serious :(

### CODE REVIEW

**Comment by jodator** [comment]
Hey, @ckeditor/qa-team - there's a new feature to check - `AutoLink`. It has its own manual test and section in Link feature guide.

cc @Reinmar since Olek is out could you take a quick look at the feature guide:

**Comment by pomek** [comment]
Test case:

1. Use [all-features](https://github.com/ckeditor/ckeditor5/blob/master/tests/manual/all-features.js) sample. Add the `AutoLink` plugin.
2. Clear the editor.
3. Type: `Type whatever.`
4. Each word is treated as link.

image

The same result I've got in the [autolink MT](http://localhost:8125/ckeditor5-link/tests/manual/autolink.html).

File: packages/ckeditor5-link/src/autolink.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+// Simplified email test - should be run over previously found URL.
+const EMAIL_REG_EXP = /^[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}$/i;
+
+/**
+ * The auto link plugin.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class AutoLink extends Plugin {
```
Comment (pomek) [inline packages/ckeditor5-link/src/autolink.js:37]: The plugin uses a few things that are provided by other plugins and as I can see, the dependencies are not specified:

* `editor.commands.get( 'enter' )` - what if missing? I know, little chance for such case but still...
* `const shiftEnterCommand = editor.commands.get( 'shiftEnter' );` - the same as above,
* `writer.setAttribute( 'linkHref', linkHrefValue, range );` - missing schema definition or `LinkEditing` which defines it.

File: packages/ckeditor5-link/src/autolink.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+const URL_REG_EXP = new RegExp(
+	// Group 1: Line start or after a space.
+	'(^|\\s)' + // Match .
+	// Group 2: Full detected URL.
+	'(' +
+		// Group 3 + 4: Protocol + domain.
+		'(([a-z]{3,9}:(?:\\/\\/)?)(?:[\\w]+)?[a-z0-9.-]+|(?:www\\.|[\\w]+)[a-z0-9.-]+)' +
+		// Group 5: Optional path + query string + location.
+		'((?:\\/[+~%/.\\w\\-_]*)?\\??(?:[-+=&;%@.\\w_]*)#?(?:[.!/\\\\\\w]*))?' +
+	')$', 'i' );
```
Comment (pomek) [inline packages/ckeditor5-link/src/autolink.js:25]: Something is wrong in the regexp - https://regex101.com/r/C9sXrU/1.

File: packages/ckeditor5-link/tests/autolink.js [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+		for ( const unsupportedURL of unsupportedURLs ) {
+			it( `should not detect "${ unsupportedURL }" as a valid URL`, () => {
+				simulateTyping( unsupportedURL + ' ' );
+
+				expect( getData( model ) ).to.equal(
+					`<paragraph>${ unsupportedURL } []</paragraph>` );
+			} );
+		}
+	} );
```
Comment (pomek) [inline packages/ckeditor5-link/tests/autolink.js:147]: I miss a test where you check what will happen after typing a text. Not a link, not an email. Just a text.

File: packages/ckeditor5-link/tests/manual/autolink.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/* globals console:false, window, document */
```
Comment (pomek) [inline packages/ckeditor5-link/tests/manual/autolink.js:6]:
```suggestion
/* globals console, window, document */
```

**Comment by pomek** [comment]
From CI:

```
Checking dependencies...
Found some issue with dependencies.
Missing devDependencies:
- @ckeditor/ckeditor5-code-block

💥 link have a dependency problem 💥
```

**Comment by pomek** [comment]
I've updated the commit/merge message.

File: packages/ckeditor5-link/tests/autolink.js (hunk 3 above)
Comment (jodator) [reply packages/ckeditor5-link/tests/autolink.js:224]: Done.

File: packages/ckeditor5-link/src/autolink.js (hunk 2 above)
Comment (jodator) [reply packages/ckeditor5-link/src/autolink.js]: Ugh - I've must fail while tuning the RegExp. I've used different one.

File: packages/ckeditor5-link/src/autolink.js (hunk 1 above)
Comment (jodator) [reply packages/ckeditor5-link/src/autolink.js:60]: I've added a tests for `Enter` and `ShiftEnter` - I've forgot about that.

The feature is build similarly to `AutoMediaEmbed` - it assumes `MediaEmbedEditing` to be present. It uses `linkHref` as `editor.execute( 'link' )` does not accept range :/

Comment (jodator) [reply packages/ckeditor5-link/src/autolink.js:60]: Added here: #7490.

**Comment by jodator** [comment]
@pomek I think I've addressed all the issues.

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  Paste a link.
2.  Press Shift + Enter.
3.  Undo.
4.  Press Enter.

## Expected result

A text is converted to a link.

## Current result

A text remains as a plain text.

## GIF

**Comment by jodator** [comment]
> A text remains as a plain text.

I think that the selection remains after `<br>` so the second enter is executed on:

```
https://google.pl<br>
[]
```

In that case it works OK.

Do you recall a bug with rendering selection after <kbd>shift</kbd>+<kbd>enter</kbd>? I think that I saw something like that.

**Comment by jodator** [comment]
@Mgsy - I've found something similar: #1576 looks like Chrome bug.

**Comment by Mgsy** [comment]
> I think that the selection remains after `<br>` so the second enter is executed on:

True, now I see that the selection doesn't change - it stays after `<br>`. Only the caret renders in incorrect place.

**Comment by jodator** [comment]
Yep, FF works fine:

**Comment by jodator** [comment]
@pomek two more things to add. I've forgot about disabling this.

**Comment by jodator** [comment]
@pomek OK I've fixed the issues with disabling `AutoLink` feature in code blocks and when using `forceDisabled()`.

**Comment by Mgsy** [comment]
## Steps to reproduce

1.  Paste a link.
2.  Press Shift + Enter.
3.  Type something.

## Current result

A user types with `linkHref` attribute.



## Expected result

A user types without `linkHref` attribute, just like after inserting a link standard way.

**Comment by jodator** [comment]
@Mgsy dup of: #7459 ?

**Comment by Mgsy** [comment]
> [**@Mgsy**](https://github.com/Mgsy) dup of: [#7459](https://github.com/ckeditor/ckeditor5/issues/7459) ?

Yes, it looks like this is the same issue :)

**Comment by Mgsy** [comment]
I've tested it in various browsers and the feature looks good :+1:

**Comment by jay2503** [comment]
is this feature released with `v20.0.0`? I could not find autolink.js under source with the latest installation.

**Comment by jodator** [comment]
@jay2503 - it will be available in the upcoming release. If everything [goes well](https://github.com/ckeditor/ckeditor5/issues/7479) it should be released next week.

**Pull request merged**